evennia/contrib/base_systems/character_creator/example_menu.py

```diff
@@ -0,0 +1,444 @@
+    """Selecting objects to start with"""
+    char = caller.new_char
+
+    # another decision, so save the resume point
+    char.db.chargen_step = "menunode_choose_objects"
+    
+    text = dedent("""\
+        |wStarting Objects|n
+        
+        Whether it's a cosmetic outfit, a starting weapon, or a professional
+        tool kit, you probably want to let your players have a choice in
+        what objects they start out with.
+        """)
+
+    help = "An overview of what the choice affects - whether it's purely aesthetic or mechanical, and whether you can change it later - are good here."
+    
+    options = []
+    
+    for proto in _EXAMPLE_PROTOTYPES:
+        # use the key as the option description, but pass the whole prototype
+        options.append({"desc": f"Choose {_INFLECT.an(proto['key'])}", "goto": ( _set_object_choice, {"proto": proto})})
+
+    options.append({"key": ("(Back)", "back", "b"), "desc": "Go back to the previous step", "goto": "menunode_multi_choice"})
+
+    return (text, help), options
+
+def _set_object_choice(caller, raw_string, proto, **kwargs):
+    """Save the selected starting object(s)"""
+    
+    # we DON'T want to actually create the object, yet! that way players can still go back and change their mind
+    # instead, we save what object was chosen - in this case, by saving the prototype dict to the character
+    caller.new_char.db.starter_weapon = proto
+    
+    # continue to the next step
+    return "menunode_choose_name"
+
+
+#########################################################
+#                Choosing a Name
+#########################################################
+
+def menunode_choose_name(caller, raw_string, **kwargs):
+    """Name selection"""
+    char = caller.new_char
+
+    # another decision, so save the resume point
+    char.db.chargen_step = "menunode_choose_name"
+    
+    text = dedent("""\
+        |wChoosing a Name|n
+        
+        Especially for roleplaying-centric games, being able to choose your
+        character's name after deciding everything else, instead of before,
+        is really useful.
+        
+        Enter a name here to check if it's available.
+        """)
+    help = "You'll have a chance to change your mind before confirming, even if the name is free."
+    # since this is a free-text field, we just have the one
+    options = [ { "key": "_default", "goto": "menunode_check_charname" } ]
+    return (text, help), options
+
+def menunode_check_charname(caller, raw_string, **kwargs):
+    """Check and confirm name choice"""
+    
+    # strip any extraneous whitespace from the raw text
+    # if you want to do any other validation on the name, e.g. no punctuation allowed, this is the place!
+    charname = raw_string.strip()
+
+    # check to make sure that the name doesn't already exist
+    candidates = Character.objects.filter_family(db_key__iexact=charname)
+    if len(candidates):
```

Comment:
Why echo the same node? Would it not make more sense to say "This was already taken", and go back to the `menunode_choose_name` node directly - that one likely contains info about the name you can pick (it could fail validation for different reasons after all, such as containing invalid characters etc - things that could be informed of in the first node). 

An alternative to doing two nodes like this would thus be to use a storage-helper (useful for all nodes in the menu) and a validation-helper (for the name node only) and have a single menu node that repeats. This is more a suggestion (if you are not familiar with the pattern) rather than a request for change though. 

```python
def _store_data(caller, raw_string, **kwargs): 
    # generic storage pass-through (useful for all nodes)
    data = kwargs.pop("data", None) 
    next_node = kwargs.pop("next_node", None)  
    if data: 
        character.ndb.chargen_data[data['key']] = data['value']
    return next_node, kwargs 

def _validate_name(caller, raw_string, **kwargs): 
    # validate input 
    if valid: 
        return None, {"name": name}
    else: 
        caller.msg("This name is invalid because ...") 
        return None, {} 

def node_name(caller, raw_string, **kwargs): 

     new_name = kwargs.get("name")

     if new_name: 
         text = "Are you sure?" 
         options = (
                {
                    "key": ("Yes", "y"), 
                    "goto": (_store_data, {"data": {"key": "name", "value": new_name}})
                 },{
                     "key" ("No", "n"),
                     "goto": "node_name"
                }
         )
     else: 
         text = "What name do you want?" 
         options = {"key": "_default",  
                          "goto": _validate_name}
    return text, options

```